### PR TITLE
Allow update to be performed by a delegate

### PR DESCRIPTION
**Pull request by febo**
This PR adds the `ProgrammableConfig` metadata delegate. The `ProgrammableConfig` is a "collection-level" delegate that can update the programmable config on the metadata account of NFTs of the same collection as long as the collection `key` of the NFT matches the collection `mint` and the update authorities of both the collection and the NFT item match.

### CODE REVIEW

File: token-metadata/program/src/state/metadata.rs [hunk 1]
```diff
@@ -98,104 +102,114 @@ impl Metadata {
             ..
         } = args;
 
-        if let Some(data) = data {
-            if !self.is_mutable {
-                return Err(MetadataError::DataIsImmutable.into());
-            }
+        // only allow update if the token standard is known
+        let token_standard = self
```
Comment (samuelvanderwaal) [inline token-metadata/program/src/state/metadata.rs:106]: Since this is an update and the update authority or delegate is signing I think we could just automatically update their token standard for them if it's not set, but that can be a separate PR.

**Review by samuelvanderwaal — APPROVED** [review]

File: token-metadata/program/src/state/metadata.rs (hunk 1 above)
Comment (febo) [reply token-metadata/program/src/state/metadata.rs]: Done.

**Review by samuelvanderwaal — DISMISSED** [review]

File: token-metadata/program/src/state/metadata.rs [hunk 2]
```diff
@@ -98,104 +103,122 @@ impl Metadata {
...
             }
-            CollectionToggle::Clear => {
-                if let Some(current_collection) = self.collection.as_ref() {
-                    // Can't change a verified collection in this command.
-                    if current_collection.verified {
-                        return Err(MetadataError::CannotUpdateVerifiedCollection.into());
+
+            // if the Collection data is 'Set', only allow updating if it's unverified
+            // or if it exactly matches the existing collection info; if the Collection data
+            // is 'Clear', then only set to 'None' it if it's unverified.
```
Comment (danenbm) [inline token-metadata/program/src/state/metadata.rs:137]: ultra nit: typo in this comment

**Review by danenbm — APPROVED** [review]
LGTM

File: token-metadata/js/test/update.test.ts [hunk 3]
```diff
@@ -1422,3 +1421,109 @@ test('Update: Cannot Update pNFT Config with locked token', async (t) => {
     /Cannot update the rule set of a programmable asset that has a delegate/i,
   );
 });
+
+test('Update: programmable config delegate update collection item', async (t) => {
```
[the remark this answers is not in the document]
Comment (febo) [reply token-metadata/js/test/update.test.ts:1425]: This is a good point, I will add this. The only caveat is whether we should fail the transaction if you try to update a field you are not allowed or the update should succeed but don't change anything.

[the remark this answers is not in the document]
Comment (danenbm) [reply token-metadata/js/test/update.test.ts:1425]: To me it seems better to fail, otherwise maybe someone tries to use the wrong delegate and thinks they were successful but there was a silent failure.  Although most people probably should and would choose to inspect the metadata after doing something to see if it really updated.

[the remark this answers is not in the document]
Comment (febo) [reply token-metadata/js/test/update.test.ts]: Done – now there is a validation of the update args.

File: token-metadata/program/src/state/metadata.rs (hunk 2 above)
Comment (febo) [reply token-metadata/program/src/state/metadata.rs]: Done.

**Review by danenbm — APPROVED** [review]

**Review by samuelvanderwaal — APPROVED** [review]
